### PR TITLE
GH-37410: [C++][Gandiva] Add support for using LLVM shared library

### LINKED ISSUE #37410: [C++] libgandiva.so links with libzstd.so even though the Gandiva module doesn't depend on zstd and zstd isn't enabled
### Describe the bug, including details regarding any error messages, version, and platform.

Here are cmake args:

-DCMAKE_CXX_STANDARD=17 -DARROW_SIMD_LEVEL=NONE -DARROW_RUNTIME_SIMD_LEVEL=NONE  -DPython3_EXECUTABLE=/usr/local/bin/python3.9 -DARROW_ACERO:BOOL=true -DARROW_WITH_BROTLI:BOOL=false -DARROW_BUILD_UTILITIES:BOOL=true -DARROW_WITH_BZ2:BOOL=false -DARROW_COMPUTE:BOOL=true -DARROW_CSV:BOOL=true -DARROW_DATASET:BOOL=true -DARROW_FILESYSTEM:BOOL=true -DARROW_FLIGHT:BOOL=false -DARROW_GANDIVA:BOOL=true -DARROW_HDFS:BOOL=true -DARROW_IPC:BOOL=true -DARROW_JSON:BOOL=true -DARROW_WITH_LZ4:BOOL=false -DARROW_MIMALLOC:BOOL=false -DARROW_ORC:BOOL=false -DARROW_PARQUET:BOOL=true -DPARQUET_REQUIRE_ENCRYPTION:BOOL=true -DARROW_S3:BOOL=true -DARROW_SKYHOOK:BOOL=false -DARROW_WITH_SNAPPY:BOOL=false -DARROW_SUBSTRAIT:BOOL=false -DARROW_TENSORFLOW:BOOL=false -DARROW_TESTING:BOOL=true -DARROW_WITH_UTF8PROC:BOOL=true -DARROW_WITH_ZLIB:BOOL=false -DARROW_WITH_ZSTD:BOOL=false -DCMAKE_C_COMPILER:STRING="cc"  -DCMAKE_CXX_COMPILER:STRING="c++"  -DCMAKE_C_FLAGS:STRING="-O2 -pipe  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing "  -DCMAKE_C_FLAGS_DEBUG:STRING="-O2 -pipe  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing "  -DCMAKE_C_FLAGS_RELEASE:STRING="-O2 -pipe  -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -DNDEBUG"  -DCMAKE_CXX_FLAGS:STRING="-O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include "  -DCMAKE_CXX_FLAGS_DEBUG:STRING="-O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include "  -DCMAKE_CXX_FLAGS_RELEASE:STRING="-O2 -pipe -fstack-protector-strong -isystem /usr/local/include -fno-strict-aliasing  -isystem /usr/local/include  -DNDEBUG"  -DCMAKE_EXE_LINKER_FLAGS:STRING="  -fstack-protector-strong "  -DCMAKE_MODULE_LINKER_FLAGS:STRING="  -fstack-protector-strong "  -DCMAKE_SHARED_LINKER_FLAGS:STRING="  -fstack-protector-strong "  -DCMAKE_INSTALL_PREFIX:PATH="/usr/local"  -DCMAKE_BUILD_TYPE:STRING="Release"  -DTHREADS_HAVE_PTHREAD_ARG:BOOL=YES  -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=YES  -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON  -DFETCHCONTENT_FULLY_DISCONNECTED:BOOL=ON -DARROW_BUILD_STATIC:BOOL=OFF -DARROW_WITH_BACKTRACE:BOOL=OFF -DARROW_BUILD_TESTS:BOOL=OFF -DARROW_BUILD_BENCHMARKS:BOOL=OFF -GNinja -DPython_ADDITIONAL_VERSIONS=3.9 -DBOOST_PYTHON_SUFFIX:STRING=39

Version: 13.0.0
FreeBSD 13.2

### Component(s)

C++

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

Gandiva always links LLVM statically. But we can use LLVM shared library instead to reduce `libgandiva.so` size.

### What changes are included in this PR?

Add `ARROW_LLVM_USE_SHARED` like other dependencies.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.

This has a backward incompatible change. LLVM shared library not static libraries is used by default. If you want to use LLVM static libraries, you need to specify `-DARROW_LLVM_USE_SHARED=OFF` or `-DARROW_DEPENDENCY_USE_SHARED=OFF` explicitly.
* Closes: #37410

### CODE REVIEW

**Comment by kou** [comment]
@github-actions java-jars -g wheel

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp

**Comment by kou** [comment]
I'll merge this in this week if nobody objects this.

**Comment by js8544** [comment]
+1. Thanks!

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -470,6 +470,16 @@ takes precedence over ccache if a storage backend is configured" ON)
                 "Rely on jemalloc shared libraries where relevant"
                 ${ARROW_DEPENDENCY_USE_SHARED})
 
+  if(MSVC)
+    # It seems that conda's llvmdev for Windows doesn't provide shared
+    # library.
+    set(ARROW_LLVM_USE_SHARED_DEFAULT OFF)
```
Comment (pitrou) [inline cpp/cmake_modules/DefineOptions.cmake:476]: Hmm, I don't think we should change this for conda. Instead, conda users on Windows will have to explicitly pass `-DARROW_LLVM_USE_SHARED=OFF`.

Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake:476]: It seems that LLVM itself doesn't support shared library on Windows: https://github.com/llvm/llvm-project/blob/main/llvm/tools/llvm-shlib/CMakeLists.txt#L14-L16

File: cpp/cmake_modules/DefineOptions.cmake [hunk 2]
```diff
@@ -470,6 +470,16 @@ takes precedence over ccache if a storage backend is configured" ON)
                 "Rely on jemalloc shared libraries where relevant"
                 ${ARROW_DEPENDENCY_USE_SHARED})
 
+  if(MSVC)
+    # It seems that conda's llvmdev for Windows doesn't provide shared
+    # library.
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake:475]:
```suggestion
    # LLVM doesn't support shared library with MSVC.
```

**Comment by kou** [comment]
@github-actions crossbow submit -g linux

**Comment by kou** [comment]
+1

**Pull request merged**